frame/support/src/traits/hooks.rs

```diff
@@ -188,6 +188,24 @@ impl OnRuntimeUpgrade for Tuple {
 
 		Ok(weight)
 	}
+
+	/// `Executive` used to directly call AllPalletsWithSystem::post_upgrade() which is now a noop.
+	/// If a runtime is running an old version of `Exeuctive`, return an error instead of silently
+	/// doing nothing which can lead to confusion. See
+	/// <https://github.com/paritytech/substrate/issues/13681>.
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+		Err("If you see this error, the implementation of `pub fn try_runtime_upgrade` in your `Executive` pallet is probably old and `pre_upgrade` checks will *not* be executed. Please update your implementation of `Executive` `pub fn try_runtime_upgrade` to match Substrate master (https://github.com/paritytech/substrate/blob/master/frame/executive/src/lib.rs).".into())
+	}
+
+	/// `Executive` used to directly call AllPalletsWithSystem::post_upgrade() which is now a noop.
```

Comment:
the impl for Tuple should not really depend from the implementation of `Executive`, which is higher level abstraction. 
if we go this direction I would just say that the default impl of `try_on_runtime_upgrade` already calls the `pre/post_upgrade`, which is the real reason here. `Executive` does not do anything specific, it just calls `try_on_runtime_upgrade`.

but I think the problem which is being addressed comes from the misuse of the trait. In my opinion these `pre/post_upgrade` funcs should invoke the inner ones, and `try_on_runtime_upgrade` should not be used with `pre/post_upgrade`.

Reply:
Good suggestion, ty